File: atlas/examples/typescript/packages/app/src/ReadingListCommand.ts
import { ReadingList } from "@atlas-example/lib/ReadingList";
import type { ReadingListItem } from "@atlas-example/lib/ReadingListItem";
import { formatISO } from "date-fns";
import { z } from "zod";

/** Validates the executable's required title argument. */
const commandTitleSchema = z
  .string()
  .trim()
  .min(1, "A book title is required.");

/**
 * Handles one reading-list command and writes its normalized result.
 */
export class ReadingListCommand {
  /**
   * Creates a command from its reusable library and process output boundary.
   *
   * @param readingList - Library behavior that owns item creation.
   * @param output - Writable process stream receiving designed user output.
   */
  public constructor(
    private readonly readingList: ReadingList,
    private readonly output: NodeJS.WritableStream,
  ) {}

  /**
   * Adds one title and reports its normalized title and stable slug.
   *
   * @param title - Raw title supplied by the user.
   * @returns The created public library item.
   */
  public execute(title: string): ReadingListItem {
    const item = this.readingList.add(commandTitleSchema.parse(title));
    const date = formatISO(new Date(), { representation: "date" });
    this.output.write(`${item.title} | ${item.slug} | ${date}\n`);
    return item;
  }
}

File: atlas/examples/kotlin/app/src/main/kotlin/dev/atlas/example/app/ReadingListCommand.kt
package dev.atlas.example.app

import com.fasterxml.jackson.databind.ObjectMapper
import dev.atlas.example.lib.ReadingList
import dev.atlas.example.lib.ReadingListItem
import org.apache.commons.lang3.StringUtils

/**
 * Handles one reading-list command and renders its normalized public library item.
 */
class ReadingListCommand(
    private val readingList: ReadingList,
    private val objectMapper: ObjectMapper
) {
    /**
     * Adds one title and returns the public library type consumed by the app.
     *
     * @param arguments Title fragments supplied after the executable name.
     * @return Created reading-list item after its JSON representation is printed.
     */
    fun execute(arguments: List<String>): ReadingListItem {
        val title = StringUtils.defaultIfBlank(arguments.joinToString(" "), "The Left Hand of Darkness")
        val item = this.readingList.add(title)
        println(this.objectMapper.writeValueAsString(item))
        return item
    }
}
